src/main/java/org/tikv/service/failsafe/CircuitBreakerImpl.java

```diff
@@ -0,0 +1,201 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.tikv.service.failsafe;
+
+import io.prometheus.client.Counter;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.tikv.common.TiConfiguration;
+
+public class CircuitBreakerImpl implements CircuitBreaker {
+  private static final Logger logger = LoggerFactory.getLogger(CircuitBreakerImpl.class);
+
+  private static final Counter CIRCUIT_BREAKER_ATTEMPT_COUNTER =
+      Counter.build()
+          .name("client_java_circuit_breaker_attempt_counter")
+          .help("client circuit breaker attempt counter.")
+          .labelNames("type")
+          .register();
+
+  private final boolean enable;
+  private final int windowInSeconds;
+  private final int errorThresholdPercentage;
+  private final int requestVolumeThreshold;
+  private final int sleepWindowInSeconds;
+  private final int attemptRequestCount;
+
+  private final AtomicLong circuitOpened = new AtomicLong(-1);
+  private final AtomicReference<Status> status = new AtomicReference<>(Status.CLOSED);
+  private final AtomicLong attemptCount = new AtomicLong(0);
+  private final AtomicLong attemptSuccessCount = new AtomicLong(0);
+
+  private final CircuitBreakerMetrics metrics;
+
+  public CircuitBreakerImpl(TiConfiguration conf) {
+    this(
+        conf.isCircuitBreakEnable(),
+        conf.getCircuitBreakAvailabilityWindowInSeconds(),
+        conf.getCircuitBreakAvailabilityErrorThresholdPercentage(),
+        conf.getCircuitBreakAvailabilityRequestVolumnThreshold(),
+        conf.getCircuitBreakSleepWindowInSeconds(),
+        conf.getCircuitBreakAttemptRequestCount());
+  }
+
+  public CircuitBreakerImpl(
+      boolean enable,
+      int windowInSeconds,
+      int errorThresholdPercentage,
+      int requestVolumeThreshold,
+      int sleepWindowInSeconds,
+      int attemptRequestCount) {
+    this.enable = enable;
+    this.windowInSeconds = windowInSeconds;
+    this.errorThresholdPercentage = errorThresholdPercentage;
+    this.requestVolumeThreshold = requestVolumeThreshold;
+    this.sleepWindowInSeconds = sleepWindowInSeconds;
+    this.attemptRequestCount = attemptRequestCount;
+    this.metrics = new CircuitBreakerMetricsImpl(windowInSeconds);
+    this.metrics.addListener(getMetricsListener());
+  }
+
+  private MetricsListener getMetricsListener() {
+    return hc -> {
+      logger.info("onNext " + hc.toString());
```

Comment:
ditto, it may periodically print useless information. I think we only need to log a warning message when the circuit breaker is open, half-open, or closed. Other messages can be debug-level.

Reply:
`info` level is used here. Users can set `warn` or `error` level in prod env.

Reply:
info level is quite normal, we should carefully print info logs as well.

Reply:
if it causes too much noise, then it should not be printed in info, or it should be printed in debug level.

Reply:
change to debug level